### PR TITLE
fix: incompatible with some react rules

### LINKED ISSUE #437: Strange behaviour of `react/jsx-curly-brace-presence`/`react/self-closing-comp`
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

2.0.5

### Link to runnable example

_No response_

### Steps to reproduce

````mdx
// test.mdx

```jsx
let a = (
  <App prop="foo">{'bar'}</App> // ❌ error  Curly braces are unnecessary here  react/jsx-curly-brace-presence
)
```

<App prop="foo">{'bar'}</App> // 😬 nothing here
````
and

````mdx
```jsx
let a = <App prop={'foo'}>bar</App> ❌ error  Curly braces are unnecessary here  react/jsx-curly-brace-presence
```

<App prop={'foo'}>bar</App> ❌ error  Curly braces are unnecessary here  react/jsx-curly-brace-presence / error  Empty components are self-closing  react/self-closing-comp
````

### Expected behavior

should report only `react/jsx-curly-brace-presence` errors

### Actual behavior

reports `react/jsx-curly-brace-presence` and `react/self-closing-comp`

### Runtime

_No response_

### Package manager

_No response_

### OS

_No response_

### Build and bundle tools

_No response_

### LINKED ISSUE #488: There are serious issues when used with the `react/self-closing-comp` / `react/jsx-sort-props` rules.
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

2.2.0

### Link to runnable example

_No response_

### Steps to reproduce

```mdx
❌ Empty components are self-closing eslint(react/self-closing-comp)
<span>test</span>

💡 Fix this react/jsx-sort-props problem
<Image alt="alt" src="src" width="315" height="100" format="svg" />
↓↓↓↓↓
<Image alt="alt" format="src" height="315" src="100" width="svg" />
```

Clone this repository [u3u/eslint-config](https://github.com/u3u/eslint-config), and input the above content in any mdx file to reproduce.

### Expected behavior

Keep consistent with `mdx/code-blocks`.

### Actual behavior

1. `<span>test</span>` Regarded as an empty tag self-closing, it will be fixed to `<span />`, directly losing the content.
2. After automatic sorting of JSX attributes, the content and attributes no longer correspond.

### Runtime

Node v20

### Package manager

pnpm

### OS

macOS

### Build and bundle tools

esbuild

### PR DESCRIPTION
**Pull request by JounQin**
### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

close #437
close #488

---

**Note**: This PR's target is `release-v2` branch, so the release workflow should be run on `release-v2` branch

cc @wooorm

### CODE REVIEW

File: packages/eslint-mdx/src/worker.ts [hunk 1]
```diff
@@ -400,8 +437,49 @@ runAsWorker(
         }
 
         function handleNode(
-          node: BlockContent | MdastLiteral | PhrasingContent,
+          node: BlockContent | Code | Literal | Paragraph | PhrasingContent,
```
[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:440]: Unfortunately, this PR still targets `release-v2` with `@types/mdast` v3.

File: packages/eslint-mdx/src/worker.ts [hunk 2]
```diff
@@ -363,33 +381,52 @@ runAsWorker(
...
         function handleChildren(
-          node: BlockContent | MdastLiteral | PhrasingContent,
+          node: BlockContent | Literal | Paragraph | PhrasingContent,
         ) {
           return 'children' in node
-            ? (node.children as Array<BlockContent | PhrasingContent>).reduce<
-                JSXElement['children']
-              >((acc, child) => {
+            ? (
+                node.children as Array<BlockContent | Literal | PhrasingContent>
```
[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:388]: Unfortunately, this PR still targets `release-v2` with `@types/mdast` v3.

File: packages/eslint-mdx/src/worker.ts [hunk 3]
```diff
@@ -363,33 +381,52 @@ runAsWorker(
         processed.add(node)
 
         function handleChildren(
-          node: BlockContent | MdastLiteral | PhrasingContent,
+          node: BlockContent | Literal | Paragraph | PhrasingContent,
```
[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:384]: For example `Code` in `jsx`:

````mdx
<div>
 ```ts
 export default 'hello world'
 ```
</div>
````

File: packages/eslint-plugin-mdx/README.md [hunk 4]
```diff
@@ -137,6 +141,62 @@ eslint . --ext js,md,mdx
...
+
+A new `MDXHeading` estree node type is exported from `eslint-mdx` which represents code blocks in `mdx` like the following:
+
+```mdx
+<div>
+
+# Here's a text gradient short code!
+
+</div>
+```
```
[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-plugin-mdx/README.md:170]: Done!

File: packages/eslint-mdx/src/worker.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wooorm) [reply packages/eslint-mdx/src/worker.ts:384]: `Code` is in `BlockContent` already

`Literal` is an *abstract* interface. It doesn’t have a `type`. It’s useful for tools that introduce new node types.
`Code` is an actual node, with a type, that *extends* `Literal`.

[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:384]: Yes, I'm using `Literal` for supporting all other possible `Literal` types.

[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:384]: I just removed the unnecessary `Code` type, thanks for pointing it out.

[the remark this answers is not in the document]
Comment (wooorm) [reply packages/eslint-mdx/src/worker.ts:384]: If you want to support arbitrary literals (and parents), use `Literal` and `Parent`?

But I don’t see it as useful. I recommend using the content types

[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:384]: I'm fixing known issues for legacy v2, your suggestion may be applied to master branch instead?

[the remark this answers is not in the document]
Comment (JounQin) [reply packages/eslint-mdx/src/worker.ts:384]: @wooorm Hi, are we good to merge this PR as-is for v2 only? I'll cherry pick the commit and adopt your review on v3 later.

[the remark this answers is not in the document]
Comment (wooorm) [reply packages/eslint-mdx/src/worker.ts:384]: sure, fine to apply it later!

**Pull request merged**